nncf/torch/graph/patterns.py

```diff
@@ -11,28 +11,105 @@
  limitations under the License.
 """
 
-from nncf.torch.graph.graph import NNCFNodeExpression as N
+import copy
 from nncf.torch.graph.version_agnostic_op_names import VersionAgnosticNames
 
-LINEAR_OPS = N('linear') | N('conv2d') | N('conv_transpose2d') | N('conv3d') | \
-             N('conv_transpose3d') | N('conv1d') | N('addmm')
+import networkx as nx
+import networkx.algorithms.isomorphism as ism
 
-RELU = N(VersionAgnosticNames.RELU) | N('hardtanh')
+NODE_COUNTER = 0
 
-BN = N('batch_norm') | N('batch_norm3d')
 
-POOLING = N('adaptive_avg_pool2d') | N('adaptive_avg_pool3d') | N('avg_pool2d') | N('avg_pool3d')
+class GraphPattern:
```

Comment:
We expect from the class the following:
1) that simple patterns can be tersely specified in code; complex patterns - perhaps more verbosely, but still straightforwardly
2) that it defines a certain syntax for specifying a custom pattern in config and can parse the corresponding string representations of the pattern

Reply:
I introduce just three operations for patterns.
Let's say that A and B are two directed graphs.

 A: (a) (b)
 B: (c) (d)

1) A + B - for all weakly connected components will add nodes from B to A and append an edge from the last node at A to the first node at B
```
                      (a)    (a)     (b)    (b)
 A + B ---->           |      |       |      |
                      (c)    (d)     (c)    (d)
```
2) A | B - add graph B to graph A as a weakly subgraph

` A  |  B ----> (a) (b) (c) (d)`

3) A * B - connect all nodes of weakly subgraphs of A with all nodes of graph B
   
```
                     ---(a)---        ---(b)---
 A * B ---->         |       |        |       |
                    (c)    (d)       (c)     (d)

```


@vshampor Could we add the same syntax to our config file?

Reply:
@aleksei-kashapov this syntax is non-intuitive - even the existing syntax is better than this. It won't be accepted unless you describe some common cases in pattern specification where they become useful. Do you find the dot file-like syntax unsuitable for some reason? It seems to me that we would have to support both the dot-like syntax for specifying basic building blocks and compelx cases, and the existing + and | syntax for grouping patterns into semantic groups and for concatenating single-entry-single-exit blocks.

Reply:
> @aleksei-kashapov this syntax is non-intuitive - even the existing syntax is better than this. It won't be accepted unless you describe some common cases in pattern specification where they become useful. Do you find the dot file-like syntax unsuitable for some reason? It seems to me that we would have to support both the dot-like syntax for specifying basic building blocks and compelx cases, and the existing + and | syntax for grouping patterns into semantic groups and for concatenating single-entry-single-exit blocks.

You can take a look at the implementation of Linear ops + Swish pattern in code:

```
FULL_PATTERN_GRAPH = LINEAR_OPS + ANY_BN_ACT_COMBO | ANY_BN_ACT_COMBO | \
                     ARITHMETIC + ANY_BN_ACT_COMBO | LINEAR_OPS_SWISH_ACTIVATION
```

Reply:
dot-like syntax looks bulky. I think that it would be sufficient for all cases just use this one